### PR TITLE
[RLlib] Make most RLlib components implementers of the new `Checkpointable` API for unified checkpointing experience.

### PR DESCRIPTION
**Pull request by sven1977**
Make most RLlib components implementers of the new `Checkpointable` API for unified checkpointing experience.
In particular, these components are:
Learner, LearnerGroup, RLModule (and MARLModule), EnvRunners (single- and multi-agent), ConnectorV2 (and pipelines).

This now allows for maximum flexibility when creating checkpoints of Algorithms and any possible subsets of component(s) of this Algorithm. Each component can define its own subcomponents, each of which will create its own subdirectory in the component's checkpoint dir.

For example, when doing ...
```
[LearnerGroup].save_to_path([dir])
```

... the resulting directory tree in [dir] will be:
```
dir/
    learner/
        rl_module/
            default_policy/
            metadata.json
            state.pkl
            class_and_ctor_args.pkl
        metadata.json
        state.pkl
        class_and_ctor_args.pkl
    metadata.json
    state.pkl
    class_and_ctor_args.pkl
```

Where
* `metadata.json` contains convenience user-readable meta information of the (sub)component saved. This information is NOT part of the state and not required to restore the respective component from disk.
* `state.pkl` contains all state information that is NOT part of any subcomponent (which has its own `state.pkl` file). For example, the optimizer state of a Learner.
* `class_and_ctor_args.pkl` stores the class of the (sub)component and its required args and kwargs to create an initial instance. This file is only needed when using the `Checkpointable.from_checkpoint([dir])` method to create a new instance (and restore its state from the checkpoint).

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a 
           method in Tune, I've added it in `doc/source/tune/api/` under the 
           corresponding `.rst` file.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by simonsays1980 — APPROVED** [review]

File: doc/source/rllib/rllib-learner.rst [hunk 1]
```diff
@@ -279,12 +279,12 @@ Getting and setting state
...
             learner_group.set_weights(weights)
 
-        Set/get the state dict of all learners through learner_group via 
+        Set/get the state dict of all learners through learner_group through
         :py:meth:`~ray.rllib.core.learner.learner_group.LearnerGroup.set_state` or 
         :py:meth:`~ray.rllib.core.learner.learner_group.LearnerGroup.get_state`. 
         This includes all states including both neural network weights, 
         and optimizer states on each learner. You can set and get the weights of 
-        the RLModule of all learners through learner_group via 
+        the RLModule of all learners through learner_group through
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-learner.rst:287]: Ah, good catch! Our vale tool doesn't like "via" that's why I changed it without noticing the duplicate "through" :|

File: doc/source/rllib/rllib-learner.rst [hunk 2]
```diff
@@ -327,25 +334,27 @@ Checkpointing
...
-        :py:meth:`~ray.rllib.core.learner.learner_group.LearnerGroup.load_state`. This includes all states including neural network weights and any
-        optimizer states. Note that since the state of all of the :py:class:`~ray.rllib.core.learner.learner.Learner` instances is identical,
-        only the states from the first :py:class:`~ray.rllib.core.learner.learner.Learner` need to be saved.
+        Checkpoint the state of all learners in the :py:class:`~ray.rllib.core.learner.learner_group.LearnerGroup`
+        through :py:meth:`~ray.rllib.core.learner.learner_group.LearnerGroup.save_to_path` and restore
+        the state of a saved :py:class:`~ray.rllib.core.learner.learner_group.LearnerGroup` through :py:meth:`~ray.rllib.core.learner.learner_group.LearnerGroup.restore_from_path`.
+        A LearnerGroup's state includes the neural network weights and all optimizer states.
+        Note that since the state of all of the :py:class:`~ray.rllib.core.learner.learner.Learner` instances is identical,
+        only the states from the first :py:class:`~ray.rllib.core.learner.learner.Learner` are saved.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-learner.rst:345]: I think you are right about this. This has been the logic from the beginning, though and was not changed in this PR. 

The question is whether the extra work and disk space (having to save all n Learners' optimizer state, which is quite a hassle) is justified by the benefits this brings (close to None, I would guess). I think the gradients and states should not be too far away on average over an optimizer's lifetime. But I could be mistaken.

File: rllib/core/rl_module/marl_module.py [hunk 3]
```diff
@@ -519,7 +458,14 @@ def __post_init__(self):
 
     def get_marl_config(self) -> "MultiAgentRLModuleConfig":
         """Returns the MultiAgentRLModuleConfig for this spec."""
-        return MultiAgentRLModuleConfig(modules=self.module_specs)
+        return MultiAgentRLModuleConfig(
+            # Only set `inference_only=True` if all single-agent specs are
```
Comment (simonsays1980) [inline rllib/core/rl_module/marl_module.py:462]: Nice!

File: rllib/core/testing/tf/bc_module.py [hunk 4]
```diff
@@ -54,30 +54,30 @@ def output_specs_inference(self) -> SpecType:
...
+    def _forward_exploration(self, batch: Dict[str, Any]) -> Dict[str, Any]:
         return self._forward_shared(batch)
 
     @override(RLModule)
-    def _forward_train(self, batch: NestedDict) -> Mapping[str, Any]:
+    def _forward_train(self, batch: Dict[str, Any]) -> Dict[str, Any]:
         return self._forward_shared(batch)
 
     @override(RLModule)
-    def get_state(self, inference_only: bool = False) -> Mapping[str, Any]:
```
Comment (simonsays1980) [inline rllib/core/testing/tf/bc_module.py:76]: How is this now running when updating from a learner module to a envrunner module?

File: rllib/env/env_runner_group.py [hunk 5]
```diff
@@ -571,33 +580,43 @@ def sync_weights(
...
+                    [COMPONENT_RL_MODULE + "/" + p for p in policies]
+                    if policies is not None
+                    else [COMPONENT_RL_MODULE]
                 )
+                # LearnerGroup has-a Learner has-a RLModule.
                 if isinstance(weights_src, LearnerGroup):
-                    weights = weights["learner_state"]["rl_module"]
+                    rl_module_state = weights_src.get_state(
+                        components=[COMPONENT_LEARNER + "/" + m for m in modules],
+                        inference_only=inference_only,
```
Comment (simonsays1980) [inline rllib/env/env_runner_group.py:594]: AH here it is. But this gets the state with the argument `inference_only` - why not leaving it in the `get_state` method of the `RLModule`s? Imo this makes it more clear to users what is happening.

File: rllib/env/single_agent_env_runner.py [hunk 6]
```diff
@@ -101,8 +108,8 @@ def __init__(self, config: AlgorithmConfig, **kwargs):
             # if the the module has target or critic networks not needed in sampling
             # or inference.
             # TODO (simon): Once we use `get_marl_module_spec` here, we can remove
-            # this line here as the function takes care of this flag.
-            module_spec.model_config_dict[INFERENCE_ONLY] = True
+            #  this line here as the function takes care of this flag.
+            module_spec.inference_only = True
```
Comment (simonsays1980) [inline rllib/env/single_agent_env_runner.py:112]: That's nicer!

File: rllib/utils/checkpoints.py [hunk 7]
```diff
@@ -497,32 +633,48 @@ def get_checkpoint_info(checkpoint: Union[str, Checkpoint]) -> Dict[str, Any]:
...
+            / COMPONENT_LEARNER_GROUP
+            / COMPONENT_LEARNER
+            / COMPONENT_RL_MODULE
+        )
+        if modules_dir.is_dir():
+            module_ids = set()
+            for module_id in modules_dir.iterdir():
+                # Only add subdirs (those are the ones where the RLModule data
+                # is stored, not files (could be json metadata files).
+                if (modules_dir / module_id).is_dir():
```
Comment (simonsays1980) [inline rllib/utils/checkpoints.py:660]: Now a non-existent case but it near future possibly used: if an `RLModule` holds other `RLModule`s which hold other`RLModule`s (e.g. for different situations different modules for each agent in a MA setup) this will not work anymore.

**Review by simonsays1980 — APPROVED** [review]
LGTM. There a couple of things that need some thoughts and adressing imo: 

1. Replay Buffers are mssing from the state.
2. Cloud synching might not be possible with this setup (possibly only via `tune`).
3. With this setup renaming components becomes dangerous b/c older versions cannot be loaded anymore
4. With this very big change in checkpointing we might want to think about keeping the old form of checkpoint loading or supply users with a "translation" method.

File: doc/source/rllib/rllib-learner.rst (hunk 1 above)
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-learner.rst:287]: fixed

File: doc/source/rllib/rllib-learner.rst [hunk 8]
```diff
@@ -293,20 +293,27 @@ Getting and setting state
 
         .. testcode::
 
+            from ray.rllib.core import COMPONENT_RL_MODULE
+
             # Get the Learner's RLModule weights and optimizer states.
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-learner.rst:298]: done

File: doc/source/rllib/rllib-learner.rst [hunk 9]
```diff
@@ -293,20 +293,27 @@ Getting and setting state
...
             state = learner.get_state()
             learner.set_state(state)
 
             # Only get the RLModule weights (as numpy arrays).
```
[the remark this answers is not in the document]
Comment (sven1977) [reply doc/source/rllib/rllib-learner.rst:302]: clarified. You get a dict of numpy arrays (as opposed to a dict of torch/tf tensors).

File: rllib/algorithms/algorithm.py [hunk 10]
```diff
@@ -2185,45 +2217,138 @@ def add_module(
...
+        if new_agent_to_module_mapping_fn is not None:
+            self.config.multi_agent(policy_mapping_fn=new_agent_to_module_mapping_fn)
+        self.config.rl_module(rl_module_spec=marl_spec)
+        if new_should_module_be_updated is not None:
+            self.config.multi_agent(policies_to_train=new_should_module_be_updated)
+        self.config.freeze()
+
+        def _add(_env_runner, _module_spec=module_spec):
+            # Add the RLModule to the existing one on the EnvRunner.
+            _env_runner.module.add_module(
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:2250]: Hmm, true. In this case, things will break, though. We should just error out and say that Modules cannot be added dynamically to an initially single-agent setup.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:2250]: Added an error to `add_module` in case user is on single-agent setup.

File: rllib/algorithms/algorithm.py [hunk 11]
```diff
@@ -2382,60 +2515,163 @@ def save_checkpoint(self, checkpoint_dir: str) -> None:
...
+        state = {}
+
+        # Get (local) EnvRunner state (w/o RLModule).
+        if self._check_component(COMPONENT_ENV_RUNNER, components, not_components):
+            state[COMPONENT_ENV_RUNNER] = self.workers.local_worker().get_state(
+                components=self._get_subcomponents(COMPONENT_RL_MODULE, components),
+                not_components=force_list(
+                    self._get_subcomponents(COMPONENT_RL_MODULE, not_components)
+                )
+                + [COMPONENT_RL_MODULE],
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:2599]: Great question! B/c we don't want the RLModule state from the EnvRunners (it's `inference_only` anyways and already provided in full by the Learners).

So we force-add "COMPONENT_RL_MODULE" to the `not_components` arg to filter it out.

[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:2599]: Added a comment to explain.

File: rllib/algorithms/algorithm.py [hunk 12]
```diff
@@ -2382,60 +2515,163 @@ def save_checkpoint(self, checkpoint_dir: str) -> None:
...
+            COMPONENT_EVAL_ENV_RUNNER, components, not_components
+        ):
+            state[
+                COMPONENT_EVAL_ENV_RUNNER
+            ] = self.evaluation_workers.local_worker().get_state(
+                components=self._get_subcomponents(COMPONENT_RL_MODULE, components),
+                not_components=force_list(
+                    self._get_subcomponents(COMPONENT_RL_MODULE, not_components)
+                )
+                + [COMPONENT_RL_MODULE],
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/algorithm.py:2614]: Added a comment to explain.

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py [hunk 13]
```diff
@@ -22,25 +21,33 @@ def setup(self):
...
+        inference_only: bool = False,
+        **kwargs,
+    ) -> StateDict:
+        state = super(PPOTorchRLModule, self).get_state(inference_only=inference_only)
         # If this module is not for inference, but the state dict is.
         # Note, for stateful modules, we need the full state dict.
-        if not self.inference_only and not self.is_stateful() and inference_only:
-            # Call the local hook to remove or rename the parameters.
-            return self._inference_only_get_state_hook(state_dict)
+        if not self.config.inference_only and not self.is_stateful() and inference_only:
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:40]: Great catch! Added the force-set:
```
            self.config.inference_only = True
```

File: rllib/algorithms/ppo/torch/ppo_torch_rl_module.py [hunk 14]
```diff
@@ -142,18 +149,6 @@ def _set_inference_only_state_dict_keys(self) -> None:
             for name in state_dict
             if "vf" in name or name.startswith("encoder.critic_encoder")
         ]
-        # Do we use a separate encoder for the actor and critic?
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/algorithms/ppo/torch/ppo_torch_rl_module.py:145]: No, I removed the renaming logic. It's super confusing to have parts of the model magically be renamed. We should simply just remove stuff from a module (e.g. the entire vf weights/biases), but never rename anything.

So this got resolved. May need some more cleanup in the future to make the `inference_only` API more unified. 

For example, all RLModules should implement an `InferenceOnly` API (similar to a TargetNetAPI or a ValueFunction API) and then override a single method that removes stuff from the network.
